python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -701,6 +702,45 @@ def cummax(self):
         """Get the column-wise cumulative maximum value in each group."""
         return self.agg("cummax")
 
+    def shift(self, periods=1, freq=None, axis=0, fill_value=None):
+        """
+        Shift each group by ``periods`` positions.
+
+        Parameters
+        ----------
+        periods : int, default 1
+            Number of periods to shift.
+        freq : str, unsupported
+        axis : 0, axis to shift
+            Shift direction. Only row-wise shift is supported
+        fill_value : optional
+            The scalar value to use for newly introduced missing values. Should
+            match the dtype of columns to fill.
+
+        Returns
+        -------
+        Series or DataFrame
+            Object shifted within each group.
+
+        Notes
+        -----
+        Parameter ``freq`` is unsupported.
+        """
+
+        if freq is not None:
+            raise NotImplementedError("Parameter freq is unsupported.")
+
+        if not axis == 0:
+            raise NotImplementedError("Only axis=0 is supported.")
+
+        data_names = [
+            x for x in self.obj._column_names if x not in self.grouping.names
+        ]
+        data = self.obj._data.select_by_label(data_names)
+        fill_values = [fill_value] * len(data)
```

Comment:
No changes suggested here. Just a note.
fill_values of libcudf seem underutilized because all of `fill_values` are of same type due to replication.

Reply:
Should we expose `fill_values` as an array instead of a single `fill_value`? Seems like the only use case for `fill_value` is `None` right now. :shrug:

Reply:
@rapidsai/cudf-python-codeowners

Reply:
Besides `None`, there's another use case - if the columns are of different numerical types and `fill_value` is a numerical value, it will try to cast to the column type before performing the fill.

Reply:
Offline sync up:
- Will support multiple `fill_values` in the API
- Gpu scalar type casting is currently left as unsupported, need to be addressed by: https://github.com/rapidsai/cudf/issues/7835